lib/dal/src/func/view.rs

```diff
@@ -1,4 +1,263 @@
+use crate::func::FuncKind;
+use crate::schema::variant::leaves::LeafInputLocation;
+use crate::{
+    ActionKind, ComponentId, DalContext, Func, FuncBackendKind, FuncBackendResponseType, FuncError,
+    FuncId, SchemaVariantError, SchemaVariantId,
+};
+
+pub mod summary;
+
+#[remain::sorted]
+#[derive(Error, Debug)]
+pub enum FuncViewError {
+    #[error("func error: {0}")]
+    Func(#[from] FuncError),
+    #[error("func argument error: {0}")]
+    FuncArgument(#[from] FuncArgumentError),
+    #[error("schema variant error: {0}")]
+    SchemaVariant(#[from] SchemaVariantError),
+}
+
+type FuncViewResult<T> = Result<T, FuncViewError>;
+
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct AttributePrototypeView {}
+
+// #[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+// #[serde(rename_all = "camelCase")]
+// pub struct ValidationPrototypeView {
+//     schema_variant_id: SchemaVariantId,
+//     prop_id: PropId,
+// }
+//
+#[remain::sorted]
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "type", rename_all = "camelCase")]
+pub enum FuncAssociations {
+    #[serde(rename_all = "camelCase")]
+    Action {
+        schema_variant_ids: Vec<SchemaVariantId>,
+        kind: ActionKind,
+    },
+    #[serde(rename_all = "camelCase")]
+    Attribute {
+        prototypes: Vec<AttributePrototypeView>,
+        arguments: Vec<FuncArgumentView>,
+    },
+    #[serde(rename_all = "camelCase")]
+    Authentication {
+        schema_variant_ids: Vec<SchemaVariantId>,
+    },
+    #[serde(rename_all = "camelCase")]
+    CodeGeneration {
+        schema_variant_ids: Vec<SchemaVariantId>,
+        component_ids: Vec<ComponentId>,
+        inputs: Vec<LeafInputLocation>,
+    },
+    #[serde(rename_all = "camelCase")]
+    Qualification {
+        schema_variant_ids: Vec<SchemaVariantId>,
+        component_ids: Vec<ComponentId>,
+        inputs: Vec<LeafInputLocation>,
+    },
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct FuncArgumentView {
+    pub id: FuncArgumentId,
+    pub name: String,
+    pub kind: FuncArgumentKind,
+    pub element_kind: Option<FuncArgumentKind>,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct FuncView {
+    pub id: FuncId,
+    pub kind: FuncKind,
+    pub name: String,
+    pub display_name: Option<String>,
+    pub description: Option<String>,
+    pub code: Option<String>,
+    pub types: String,
+    pub is_builtin: bool,
+    pub is_revertible: bool,
+    pub associations: Option<FuncAssociations>,
+}
+
+impl FuncView {
+    pub async fn assemble(ctx: &DalContext, func: &Func) -> FuncViewResult<Self> {
+        let arguments = FuncArgument::list_for_func(ctx, func.id).await?;
+
+        let (_associations, _input_type) = match func.kind {
```

Comment:
Work in progress for new associations detection. This will likely be the starting point for the next PR.